[WebXR] Implement XRWebGLBinding constructor

svillar opened the pull request:
#### 560926efcc3867ba87c1f064a32874b4999ed2ee
<pre>
[WebXR] Implement XRWebGLBinding constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=301102">https://bugs.webkit.org/show_bug.cgi?id=301102</a>

Reviewed by Dan Glastonbury.

Adding the implementation of the constructor of the bindings
object as per the WebXR spec. So far is basically a sequence
of checks.

Replaced subtests expectations from FAIL to PASS and unskipping
the WPT test that verifies the constructor behaviour.

* LayoutTests/imported/w3c/web-platform-tests/webxr/layers/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/layers/xrWebGLBinding_constructor.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
(WebCore::XRWebGLBinding::create):
(WebCore::XRWebGLBinding::XRWebGLBinding):
* Source/WebCore/Modules/webxr/XRWebGLBinding.h:

Canonical link: <a href="https://commits.webkit.org/301937@main">https://commits.webkit.org/301937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9ecd47381e5dde9f764f1aec2c798961fbc54ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127360 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134424 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78915 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129232 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55531 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96929 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64923 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130308 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114071 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77423 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36957 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32186 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77806 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119397 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136907 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125823 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54019 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41627 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105449 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54530 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110420 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105126 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26835 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50647 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29095 "Found 1 new test failure: http/tests/misc/ftp-eplf-directory.py (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51598 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53956 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60043 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158861 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53189 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39722 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
```diff
@@ -29,15 +29,44 @@
+ExceptionOr<Ref<XRWebGLBinding>> XRWebGLBinding::create(Ref<WebXRSession>&& session, WebXRWebGLRenderingContext&& context)
+{
+    if (session->ended())
+        return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding with a XRSession that has ended."_s };
```
"... create an XRWebGLBinding with an XRSession ..."

djg commented on Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
```diff
@@ -29,15 +29,44 @@
+    return WTF::switchOn(context,
+        [&](const RefPtr<WebGLRenderingContextBase>& baseContext) -> ExceptionOr<Ref<XRWebGLBinding>> {
+            if (!baseContext || baseContext->isContextLost())
+                return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding with a lost WebGL context."_s };
```
ditto

djg commented on Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
```diff
@@ -29,15 +29,44 @@
+                return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding with a lost WebGL context."_s };
+
+            if (!isImmersive(session->mode()))
+                return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding for non immersive sessions."_s };
```
ditto

djg commented on Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
```diff
@@ -29,15 +29,44 @@
+                return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding for non immersive sessions."_s };
+
+            if (!baseContext->isXRCompatible())
+                return Exception { ExceptionCode::InvalidStateError, "Cannot create a XRWebGLBinding with a non XR compatible WebGL context."_s };
```
ditto

djg commented on LayoutTests/imported/w3c/web-platform-tests/webxr/layers/xrWebGLBinding_constructor.https-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
-FAIL Ensure that XRWebGLBinding's constructor throws appropriate errors using webgl assert_equals: Should get InvalidStateError for creating with inline session. expected "InvalidStateError" but got "ReferenceError"
-FAIL Ensure that XRWebGLBinding's constructor throws appropriate errors using webgl2 assert_equals: Should get InvalidStateError for creating with inline session. expected "InvalidStateError" but got "ReferenceError"
+PASS Ensure that XRWebGLBinding's constructor throws appropriate errors using webgl
+PASS Ensure that XRWebGLBinding's constructor throws appropriate errors using webgl2
```
Are these changes going to cause testing on Cocoa platforms to fail?  If so, we need to split expectations for Cocoa platforms and GTK/WPE.

svillar replied:
Cocoa has all these tests skipped AFAIK

djg reviewed (approved):
:shipit:

svillar committed:
[WebXR] Implement XRWebGLBinding constructor
https://bugs.webkit.org/show_bug.cgi?id=301102

Reviewed by Dan Glastonbury.

Adding the implementation of the constructor of the bindings
object as per the WebXR spec. So far is basically a sequence
of checks.

Replaced subtests expectations from FAIL to PASS and unskipping
the WPT test that verifies the constructor behaviour.

* LayoutTests/imported/w3c/web-platform-tests/webxr/layers/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/layers/xrWebGLBinding_constructor.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webxr/XRWebGLBinding.cpp:
(WebCore::XRWebGLBinding::create):
(WebCore::XRWebGLBinding::XRWebGLBinding):
* Source/WebCore/Modules/webxr/XRWebGLBinding.h:

Canonical link: https://commits.webkit.org/301937@main
